### PR TITLE
Add width and height arguments to `@render.plot`

### PR DESCRIPTION
**Pull request by jcheng5**
The primary goal of this PR is to provide an easy way to influence the width/height of the `output_plot` that's _implicitly_ generated by displaying a `@render.plot`. There will also be a separate `@output_args` decorator that will let you set any argument on any implicit output, but the specific case of plot width/height seems like such a common case that it's worth the effort to make it dead simple.

Two additional effects of this PR: the pixel dimensions of the image generated on the server side can now be influenced by `@render.plot(width=float, height=float)`, or if that is not specified, by using matplotlib/plotnine to set the Figure size (in inches). And it is now possible to set the DPI of a figure using matplotlib/plotnine, which changes the size of text.

### CODE REVIEW

File: shiny/render/_render.py [hunk 1]
```diff
@@ -217,6 +257,8 @@ def plot(
     _fn: PlotTransformer.ValueFn | None = None,
     *,
     alt: Optional[str] = None,
+    width: Optional[float] = None,
+    height: Optional[float] = None,
```
Comment (jcheng5) [inline shiny/render/_render.py:261]: TODO: Add docstring entries for these

File: shiny/render/_render.py [hunk 2]
```diff
@@ -96,35 +101,75 @@ def text(
 # ======================================================================================
 # RenderPlot
 # ======================================================================================
+
+
+def _default_plot_ui(
+    id: str, *args: Any, _params: dict[str, Any] | None = None, **kwargs: Any
```
Comment (wch) [inline shiny/render/_render.py:107]: Use `object` instead of `Any` if possible.

File: shiny/render/_render.py [hunk 3]
```diff
@@ -96,35 +101,75 @@ def text(
...
+            )
+
+        if "height" in _params:
+            extra_args["height"] = (
+                f"{_params['height']}px"
+                if isinstance(_params["height"], (int, float))
+                else str(_params["height"])
+            )
+
+        kwargs.update(extra_args)
```
Comment (wch) [inline shiny/render/_render.py:129]: Reverse the order here, so that `kwargs` wins.

File: shiny/render/_render.py [hunk 4]
```diff
@@ -96,35 +101,75 @@ def text(
...
+    # the decorator and apply them to the output_plot call.
+    if _params is not None:
+        extra_args = {}
+
+        if "width" in _params:
+            extra_args["width"] = (
+                f"{_params['width']}px"
+                if isinstance(_params["width"], (int, float))
+                else str(_params["width"])
+            )
```
Comment (wch) [inline shiny/render/_render.py:120]:
```suggestion
            extra_args["width"] = _params["width"]
```

Comment (wch) [reply shiny/render/_render.py:120]: Same for height below.

File: shiny/render/_render.py [hunk 5]
```diff
@@ -96,35 +101,75 @@ def text(
...
-        float, inputs[ResolvedId(f".clientdata_output_{name}_height")]()
+
+    # Do NOT call this unless you actually are going to respect the container dimension
+    # you're asking for. It takes a reactive dependency. If the client hasn't reported
+    # the requested dimension, you'll get a SilentException.
+    def container_size(dimension: Literal["width", "height"]) -> float:
+        result = inputs[ResolvedId(f".clientdata_output_{name}_{dimension}")]()
+        return typing.cast(float, result)
+
+    plot_size_info = PlotSizeInfo(
```
Comment (wch) [inline shiny/render/_render.py:166]: Please add comment explaining that we don't have enough information at this point to be sure of size, so we're going to build up that information.

File: shiny/render/_render.py [hunk 6]
```diff
@@ -227,6 +267,12 @@ def plot(
     alt
         Alternative text for the image if it cannot be displayed or viewed (i.e., the
         user uses a screen reader).
+    width
+        Width of the plot in pixels. If ``None``, the width will be determined by the
+        size of the corresponding :func:`~shiny.ui.output_plot`.
+    height
+        Height of the plot in pixels. If ``None``, the height will be determined by the
+        size of the corresponding :func:`~shiny.ui.output_plot`.
```
Comment (wch) [inline shiny/render/_render.py:275]: Tell people they shouldn't use this in traditional shiny apps.

File: shiny/render/_try_render_plot.py [hunk 7]
```diff
@@ -15,15 +15,113 @@
...
+        Args:
+            container_size_px_fn: A tuple of two functions that return the width and
+                height of the container, in pixels. If the user specified an explicit
+                width/height on the @render.plot decorator, then these functions should
+                not be called, as doing so will take a reactive dependency on that
+                dimension.
+            user_specified_size_px: A tuple of two floats, or None. If the user
+                specified an explicit width/height on the @render.plot decorator, then
+                the corresponding float will be that value. Otherwise, it will be None.
+            pixelratio: The device pixel ratio that was detected from the client.
```
Comment (wch) [inline shiny/render/_try_render_plot.py:39]: Change format to something like:

```
Params
-------
```

File: shiny/render/_try_render_plot.py [hunk 8]
```diff
@@ -15,15 +15,113 @@
...
+        self.pixelratio = pixelratio
+
+    def get_img_size_px(
+        self,
+        fig_initial_size_inches: tuple[float, float],
+        fig_result_size_inches: tuple[float, float],
+        dpi: float,
+    ) -> tuple[float, float, str, str]:
+        """
+        Determines the desired size of the image in logical pixels (not doubling
```
Comment (wch) [inline shiny/render/_try_render_plot.py:52]:
```suggestion
        Determines the desired size of the image in CSS pixels (not doubling
```

File: shiny/render/_try_render_plot.py [hunk 9]
```diff
@@ -15,15 +15,113 @@
...
+        resolution for retina displays--the caller needs to worry about that), and the
+        width and height attributes that should be set on the <img> tag.
+
+        Args:
+            fig_initial_size_inches: The size of the figure before the user's plotting
```
Comment (wch) [inline shiny/render/_try_render_plot.py:57]: Please make clear that "figure" refers to a matplotlib figure, and not img tag in the DOM.

File: shiny/render/transformer/_transformer.py [hunk 10]
```diff
@@ -319,14 +319,21 @@ async def _run(self) -> OT:
...
 
     def tagify(self) -> TagList | Tag | MetadataNode | str:
         if self.default_ui is None:
             raise TypeError("No default UI exists for this type of render function")
-        return self.default_ui(self.__name__)
+        return self._render_default()
+
+    def _render_default(self) -> TagList | Tag | MetadataNode | str:
+        if self.default_ui is None:
+            return TagList()
```
Comment (wch) [inline shiny/render/transformer/_transformer.py:331]: Raise error here.

File: shiny/render/transformer/_transformer.py [hunk 11]
```diff
@@ -319,14 +319,21 @@ async def _run(self) -> OT:
     def _repr_html_(self) -> str | None:
         import htmltools
 
-        if self.default_ui is None:
-            return None
```
Comment (wch) [inline shiny/render/transformer/_transformer.py:323]: Restore this code.

File: tests/e2e/plot-sizing/app.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        [ui.div(x, class_="border mb-3") for x in plot_outputs],
+    )
+
+
+@module.server
+def plot_server(
+    input: Inputs,
+    output: Outputs,
+    session: Session,
+    plot_fn: Callable[[tuple[float, float] | None], Any],
```
Comment (wch) [inline tests/e2e/plot-sizing/app.py:40]:
```suggestion
    plot_fn: Callable[[tuple[float, float] | None], object],
```

File: tests/e2e/plot-sizing/app.py [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+            p = p + theme(
+                figure_size=(fig_size[0] / dpi, fig_size[1] / dpi),
+                plot_background=element_rect(fill="lavender"),
+                dpi=dpi,
+            )
+        return p
+
+    def plot_with_pil(fig_size: tuple[float, float] | None):
+        req(fig_size is None)
+        return Image.open(Path(__file__).parent / "IMG_0997.jpg")
```
Comment (wch) [inline tests/e2e/plot-sizing/app.py:124]:
```suggestion
        return Image.open(Path(__file__).parent / "bike.jpg")
```

**Review by wch — CHANGES_REQUESTED** [review]

File: shiny/render/_try_render_plot.py (hunk 8 above)
Comment (jcheng5) [reply shiny/render/_try_render_plot.py:52]: I googled "CSS pixels" and the first hit said CSS defines pixels to be 1/96 inch. So I guess that's where the 96 came from.

File: shiny/render/_render.py (hunk 3 above)
Comment (jcheng5) [reply shiny/render/_render.py]: I couldn't reverse the order without upsetting the type checker, but I filtered out the extra_args that would conflict.

**Review by wch — APPROVED** [review]

**Comment by wch** [comment]
I just merged -- but note that we should have a changelog entry for this PR.

**Pull request merged**